File: detector.py
import cv2
import numpy as np

# 定义常量
RED = 'red'
BLUE = 'blue'
ARMOR_TYPE_STR = ['INVALID', 'SMALL', 'LARGE']

class Light:
    def __init__(self, rect):
        self.bounding_rect = rect
        self.top = (rect[0][0], rect[0][1])  # 记录灯的上点
        self.bottom = (rect[1][0], rect[1][1])  # 记录灯的下点
        self.center = np.array([(rect[0][0] + rect[1][0]) / 2, (rect[0][1] + rect[1][1]) / 2])
        self.width = rect[1][0] - rect[0][0]
        self.length = rect[1][1] - rect[0][1]
        self.color = None  # 灯的颜色

    def calculate_tilt_angle(self):
        """计算倾斜角度"""
        # cv2.minAreaRect 返回的第三个元素是旋转角度
        _, _, angle = cv2.minAreaRect(np.array(self.bounding_rect))
        # 如果角度大于 0，则调整为负数表示逆时针方向
        return -angle if angle > 0 else angle


class Armor:
    def __init__(self, light1, light2):
        self.left_light  = light1  # 左灯
        self.right_light = light2  # 右灯
        self.type = None           # 装甲板类型
        self.classfication_result = ""  # 分类结果
        self.number_img = None     # 数字图像


class Detector:
    """
    装甲板识别类
    @params: bin_thres: 二值化阈值
    @params: detect_color: 识别的颜色
    @params: l: 灯条参数
    @params: a: 装甲板参数
    """
    def __init__(self, bin_thres, color, light_params, armor_params):
        self.binary_thres = bin_thres
        self.detect_color = color
        self.l = light_params
        self.a = armor_params
        self.armors_ = []
        self.lights_ = []

    def detect(self, input_img):
        """识别装甲板"""
        binary_img = self.preprocess_image(input_img)           # 预处理图像
        self.lights_ = self.find_lights(input_img, binary_img)  # 查找灯条
        self.armors_ = self.match_lights(self.lights_)          # 匹配灯条

        if self.armors_:
            # 这里可以添加数字提取和分类的逻辑
            pass

        return self.armors_

    def preprocess_image(self, rgb_img):
        """预处理图像"""
        gray_img = cv2.cvtColor(rgb_img, cv2.COLOR_BGR2GRAY)  # 转换为灰度图像
        _, binary_img = cv2.threshold(gray_img, self.binary_thres, 255, cv2.THRESH_BINARY)  # 二值化
        return binary_img

    def find_lights(self, rgb_img, binary_img):
        """查找灯条"""
        contours, _ = cv2.findContours(binary_img, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)  # 查找轮廓
        lights = []

        for contour in contours:
            if len(contour) < 5:
                continue

            r_rect = cv2.minAreaRect(contour)  # 获取最小外接矩形
            # 确保 r_rect 是一个有效的轮廓点集
            r_rect_points = np.array(cv2.boxPoints(r_rect)).astype(int)

            light = Light(r_rect_points)

            if self.is_light(light):
                lights.append(light)

        return lights

    def is_light(self, light):
        """判断是否为灯条"""
        ratio = light.width / light.length  # 计算长宽比
        ratio_ok = self.l['min_ratio'] < ratio < self.l['max_ratio']
        angle_ok = light.calculate_tilt_angle() < self.l['max_angle']  # 调用新方法获取角度
        
        return ratio_ok and angle_ok

    def match_lights(self, lights):
        """匹配灯条"""
        armors = []
        
        for i, light_1 in enumerate(lights):
            for light_2 in lights[i+1:]:
                if light_1.color != self.detect_color or light_2.color != self.detect_color:
                    continue

                if self.contain_light(light_1, light_2, lights):
                    continue

                armor_type = self.is_armor(light_1, light_2)
                if armor_type != 'INVALID':
                    armor = Armor(light_1, light_2)
                    armor.type = armor_type
                    armors.append(armor)

        return armors

    def contain_light(self, light1, light2, lights):
        """判断灯条是否包含其他灯条"""
        points = [light1.top, light1.bottom, light2.top, light2.bottom]
        bounding_rect = cv2.boundingRect(np.array(points))  # 计算边界框

        for test_light in lights:
            if test_light.center == light1.center or test_light.center == light2.center:
                continue

            if (bounding_rect[0] <= test_light.top[0] <= bounding_rect[0] + bounding_rect[2] and
                bounding_rect[1] <= test_light.top[1] <= bounding_rect[1] + bounding_rect[3]):
                return True

        return False

    def is_armor(self, light1, light2):
        """判断是否为装甲板"""
        light_length_ratio = min(light1.length, light2.length) / max(light1.length, light2.length)
        light_ratio_ok = light_length_ratio > self.a['min_light_ratio']

        avg_light_length = (light1.length + light2.length) / 2
        center_distance = np.linalg.norm(light1.center - light2.center) / avg_light_length
        center_distance_ok = (self.a['min_small_center_distance'] <= center_distance < self.a['max_small_center_distance'] or
                              self.a['min_large_center_distance'] <= center_distance < self.a['max_large_center_distance'])

        angle = np.degrees(np.arctan2(light1.center[1] - light2.center[1], light1.center[0] - light2.center[0]))
        angle_ok = abs(angle) < self.a['max_angle']

        is_armor = light_ratio_ok and center_distance_ok and angle_ok

        return 'LARGE' if is_armor and center_distance > self.a['min_large_center_distance'] else 'SMALL' if is_armor else 'INVALID'

    def draw_results(self, img):
        """绘制结果"""
        for light in self.lights_:
            cv2.circle(img, light.top, 3, (255, 255, 255), 1)
            cv2.circle(img, light.bottom, 3, (255, 255, 255), 1)
            line_color = (255, 255, 0) if light.color == RED else (255, 0, 255)
            cv2.line(img, light.top, light.bottom, line_color, 1)

        for armor in self.armors_:
            cv2.line(img, armor.left_light.top, armor.right_light.bottom, (0, 255, 0), 2)
            cv2.line(img, armor.left_light.bottom, armor.right_light.top, (0, 255, 0), 2)

        for armor in self.armors_:
            cv2.putText(img, armor.classfication_result, armor.left_light.top, cv2.FONT_HERSHEY_SIMPLEX, 0.8, (0, 255, 255), 2)
        
        return img

# 使用示例
# light_params 和 armor_params 应根据实际需求设置
light_params = {'min_ratio': 0.5, 'max_ratio': 2.0, 'max_angle': 15}
armor_params = {'min_light_ratio': 0.5, 'min_small_center_distance': 30, 'max_small_center_distance': 100,
                'min_large_center_distance': 100, 'max_large_center_distance': 300, 'max_angle': 15}

detector = Detector(bin_thres=128, color=RED, light_params=light_params, armor_params=armor_params)

# 读取图像并进行检测
input_image = cv2.imread('2.jpg')
#input_image = cv2.imread('combine.png')
detected_armors = detector.detect(input_image)

# 绘制检测结果
out = detector.draw_results(input_image)

cv2.imshow('Detection Results', out)
cv2.waitKey(0)
cv2.destroyAllWindows()

File: detector.cpp
2// 版权所有 (c) 2022 ChenJun
// 采用MIT许可证。

// 引入OpenCV相关头文件
#include <opencv2/core.hpp>
#include <opencv2/core/base.hpp>
#include <opencv2/core/mat.hpp>
#include <opencv2/core/types.hpp>
#include <opencv2/imgproc.hpp>

// 引入标准库相关头文件
#include <algorithm>
#include <cmath>
#include <vector>

// 引入自定义消息类型头文件
#include "armor_detector/detector.hpp"
#include "auto_aim_interfaces/msg/debug_armor.hpp"
#include "auto_aim_interfaces/msg/debug_light.hpp"

// 命名空间rm_auto_aim开始
namespace rm_auto_aim
{
// Detector类的构造函数
Detector::Detector(
  const int & bin_thres, const int & color, const LightParams & l, const ArmorParams & a)
: binary_thres(bin_thres), detect_color(color), l(l), a(a)
{
}
/**
 * @brief 主函数, 检测装甲板并返回装甲板列表
 * 
 * @return 返回装甲板列表
 * @note 函数执行顺序为：预处理图像 -> 寻找灯源 -> 匹配灯源 -> 提取数字 -> 分类数字 -> 返回装甲板列表
 * @note 函数返回装甲板列表后，会根据装甲板的类型进行分类，并计算装甲板的中心坐标和旋转角度
 */
std::vector<Armor> Detector::detect(const cv::Mat & input)
{
  binary_img = preprocessImage(input);      // 预处理图像并获取二值图像
  lights_ = findLights(input, binary_img);  // 在图像中查找灯源
  armors_ = matchLights(lights_);           // 匹配并组合灯源，形成可能的装甲板

  if (!armors_.empty()) {  // 如果检测到装甲板，则进一步提取和分类数字
    classifier->extractNumbers(input, armors_);
    classifier->classify(armors_);
  }

  return armors_;  // 返回检测到的装甲板列表
}

/**
 * 预处理输入的RGB图像，转换为二值图像
 * 
 * @param rgb_img RGB图像
 * @return 返回一个二值图像
 */
cv::Mat Detector::preprocessImage(const cv::Mat & rgb_img)
{
  cv::Mat gray_img;  // 将RGB图像转换为灰度图像
  cv::cvtColor(rgb_img, gray_img, cv::COLOR_RGB2GRAY);

  cv::Mat binary_img;  // 根据二值化阈值将灰度图像转换为二值图像
  cv::threshold(gray_img, binary_img, binary_thres, 255, cv::THRESH_BINARY);

  return binary_img;
}

/**
 * 在给定的RGB图像和二值图像中查找灯源并返回一个二值图像
 * 
 * @param rgb_img RGB图像
 * @param binary_img 二值图像
 * @return 返回一个二值图像
 */
std::vector<Light> Detector::findLights(const cv::Mat & rbg_img, const cv::Mat & binary_img)
{
  using std::vector;
  vector<vector<cv::Point>> contours;
  vector<cv::Vec4i> hierarchy;
  // 在二值图像中查找轮廓
  cv::findContours(binary_img, contours, hierarchy, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_SIMPLE);

  vector<Light> lights;
  this->debug_lights.data.clear();

  // 遍历所有轮廓，寻找符合条件的灯源
  for (const auto & contour : contours) {
    if (contour.size() < 5) continue;

    auto r_rect = cv::minAreaRect(contour);
    auto light = Light(r_rect);

    if (isLight(light)) {
      auto rect = light.boundingRect();
      // 确保矩形在图像范围内
      if (  // 避免断言失败
        0 <= rect.x && 0 <= rect.width && rect.x + rect.width <= rbg_img.cols && 0 <= rect.y &&
        0 <= rect.height && rect.y + rect.height <= rbg_img.rows) {
        int sum_r = 0, sum_b = 0;
        auto roi = rbg_img(rect);
        // 遍历ROI内的每个像素点，计算红色和蓝色像素值之和
        for (int i = 0; i < roi.rows; i++) {
          for (int j = 0; j < roi.cols; j++) {
            if (cv::pointPolygonTest(contour, cv::Point2f(j + rect.x, i + rect.y), false) >= 0) {
              // 如果点在轮廓内
              sum_r += roi.at<cv::Vec3b>(i, j)[0];
              sum_b += roi.at<cv::Vec3b>(i, j)[2];
            }
          }
        }
        // 红色像素值之和大于蓝色像素值之和，则认为是红色灯源，否则为蓝色灯源
        light.color = sum_r > sum_b ? RED : BLUE;
        lights.emplace_back(light);
      }
    }
  }

  return lights;
}

/**
 * 判断给定的灯条对象是否符合识别为灯条的标准
 * 主要通过灯条对象的长宽比和倾斜角度来判断
 * 
 * @param light 灯条对象引用
 * @return 返回一个布尔值，表示该灯条对象是否被识别为灯条
 */
bool Detector::isLight(const Light & light)
{
  // 计算灯条的长宽比（短边 / 长边）
  float ratio = light.width / light.length;
  bool ratio_ok = l.min_ratio < ratio && ratio < l.max_ratio;
  bool angle_ok = light.tilt_angle < l.max_angle;
  bool is_light = ratio_ok && angle_ok;

  // 填充调试信息
  auto_aim_interfaces::msg::DebugLight light_data;
  light_data.center_x = light.center.x;
  light_data.ratio = ratio;
  light_data.angle = light.tilt_angle;
  light_data.is_light = is_light;
  this->debug_lights.data.emplace_back(light_data);

  return is_light;
}

/**
 * 根据灯条列表匹配装甲板对象
 * 
 * @param lights 灯条列表
 * @return 返回一个装甲板对象列表
 */
std::vector<Armor> Detector::matchLights(const std::vector<Light> & lights)
{
  std::vector<Armor> armors;
  this->debug_armors.data.clear();

  // 遍历所有可能的灯条配对组合
  for (auto light_1 = lights.begin(); light_1 != lights.end(); light_1++) {
    for (auto light_2 = light_1 + 1; light_2 != lights.end(); light_2++) {
      if (light_1->color != detect_color || light_2->color != detect_color) continue;

      if (containLight(*light_1, *light_2, lights)) {
        continue;
      }

      auto type = isArmor(*light_1, *light_2);
      if (type != ArmorType::INVALID) {
        auto armor = Armor(*light_1, *light_2);
        armor.type = type;
        armors.emplace_back(armor);
      }
    }
  }

  return armors;
}

/**
 * 检查边界中是否有其他灯由两个灯形成
 * 
 * @param light_1 第一个灯对象
 * @param light_2 第二个灯对象
 * @param lights 所有灯对象列表
 * @return 返回一个布尔值，表示边界中是否有其他灯
 */
bool Detector::containLight(
  const Light & light_1, const Light & light_2, const std::vector<Light> & lights)
{
  // 收集两个灯的四个点形成点集，并计算边界框
  auto points = std::vector<cv::Point2f>{light_1.top, light_1.bottom, light_2.top, light_2.bottom};
  auto bounding_rect = cv::boundingRect(points);

  // 遍历所有灯，检查是否在边界框内
  for (const auto & test_light : lights) {
    if (test_light.center == light_1.center || test_light.center == light_2.center) continue;

    if (
      bounding_rect.contains(test_light.top) || bounding_rect.contains(test_light.bottom) ||
      bounding_rect.contains(test_light.center)) {
      return true;
    }
  }

  return false;
}

/**
 * 判断两个灯是否构成装甲板，并确定装甲板类型
 * 
 * @param light_1 第一个灯对象
 * @param light_2 第二个灯对象
 * @return 返回装甲板类型（大、小或无效）
 */
ArmorType Detector::isArmor(const Light & light_1, const Light & light_2)
{
  // 计算两个灯长度比（短边/长边）
  float light_length_ratio = light_1.length < light_2.length ? light_1.length / light_2.length
                                                             : light_2.length / light_1.length;
  bool light_ratio_ok = light_length_ratio > a.min_light_ratio;

  // 计算两个灯中心距离（单位：灯的平均长度）
  float avg_light_length = (light_1.length + light_2.length) / 2;
  float center_distance = cv::norm(light_1.center - light_2.center) / avg_light_length;
  bool center_distance_ok = (a.min_small_center_distance <= center_distance &&
                             center_distance < a.max_small_center_distance) ||
                            (a.min_large_center_distance <= center_distance &&
                             center_distance < a.max_large_center_distance);

  // 计算两灯中心连线的角度
  cv::Point2f diff = light_1.center - light_2.center;
  float angle = std::abs(std::atan(diff.y / diff.x)) / CV_PI * 180;
  bool angle_ok = angle < a.max_angle;

  // 判断是否为装甲板
  bool is_armor = light_ratio_ok && center_distance_ok && angle_ok;

  // 判断装甲板类型
  ArmorType type;
  if (is_armor) {
    type = center_distance > a.min_large_center_distance ? ArmorType::LARGE : ArmorType::SMALL;
  } else {
    type = ArmorType::INVALID;
  }

  // 填充调试信息
  auto_aim_interfaces::msg::DebugArmor armor_data;
  armor_data.type = ARMOR_TYPE_STR[static_cast<int>(type)];
  armor_data.center_x = (light_1.center.x + light_2.center.x) / 2;
  armor_data.light_ratio = light_length_ratio;
  armor_data.center_distance = center_distance;
  armor_data.angle = angle;
  this->debug_armors.data.emplace_back(armor_data);

  return type;
}

/**
 * 获取所有装甲板的数字图像
 * 
 * @return 返回一个OpenCV的Mat对象，包含所有装甲板的数字图像, 如果装甲板列表为空，则返回一个20x28大小的单通道灰度图
 */
cv::Mat Detector::getAllNumbersImage()
{
  if (armors_.empty()) {
    return cv::Mat(cv::Size(20, 28), CV_8UC1);
  } else {
    std::vector<cv::Mat> number_imgs;
    number_imgs.reserve(armors_.size());
    for (auto & armor : armors_) {
      number_imgs.emplace_back(armor.number_img);
    }
    cv::Mat all_num_img;
    cv::vconcat(number_imgs, all_num_img);
    return all_num_img;
  }
}

/**
 * 绘制检测结果
 * 
 * @param img 输入图像
 */
void Detector::drawResults(cv::Mat & img)
{
  // 绘制灯
  for (const auto & light : lights_) {
    cv::circle(img, light.top, 3, cv::Scalar(255, 255, 255), 1);
    cv::circle(img, light.bottom, 3, cv::Scalar(255, 255, 255), 1);
    auto line_color = light.color == RED ? cv::Scalar(255, 255, 0) : cv::Scalar(255, 0, 255);
    cv::line(img, light.top, light.bottom, line_color, 1);
  }

  // 绘制装甲板
  for (const auto & armor : armors_) {
    cv::line(img, armor.left_light.top, armor.right_light.bottom, cv::Scalar(0, 255, 0), 2);
    cv::line(img, armor.left_light.bottom, armor.right_light.top, cv::Scalar(0, 255, 0), 2);
  }

  // 显示数字及其置信度
  for (const auto & armor : armors_) {
    cv::putText(
      img, armor.classfication_result, armor.left_light.top, cv::FONT_HERSHEY_SIMPLEX, 0.8,
      cv::Scalar(0, 255, 255), 2);
  }
}

}  // namespace rm_auto_aim
